Add platform option to crane copy and digest

dsbrng25b opened the pull request:
Add platform option to crane copy and digest. If the remote reference is an index/list it only copies the image which matches the specified platform.

This would be a possible solution for #741

jonjohnsonjr commented on pkg/crane/get.go:
```diff
@@ -41,5 +42,46 @@ func getManifest(r string, opt ...Option) (*remote.Descriptor, error) {
+// referece points to a index and the platform option is set, instead of the
+// index descriptor, the descriptor of the image which matches the platform is
+// returned.
+func getRemoteDescriptor(ref name.Reference, o options) (*remote.Descriptor, error) {
```
Rather than doing this logic here, it might be easier to just do something in `pkg/crane` like:

```go
if opt.platform != nil {
  img, err := desc.Image()
  ...
}
```

in copy and manifest --  I think that would be a smaller change.

dsbrng25b replied:
I don't get what you mean. The logic (for loop below) is required to find the matching platform if the descriptor is a index/list.

jonjohnsonjr replied:
I'd like to avoid duplicating this logic too much -- this happens in `remote`: https://github.com/google/go-containerregistry/blob/72597da50ce903895087828b5fc4b918636366f4/pkg/v1/remote/index.go#L125-L159

Your loop here assumes there's only ever one level of nesting within an image index, which is usually true, but it's possible to have a [nested index](https://github.com/google/go-containerregistry/tree/72597da50ce903895087828b5fc4b918636366f4/pkg/v1/remote#anatomy-of-an-index), which this wouldn't handle.

If you make `crane.WithPlatform` additionally pass the option down through to `remote.WithPlatform` (similar to [this](https://github.com/google/go-containerregistry/blob/72597da50ce903895087828b5fc4b918636366f4/pkg/crane/options.go#L49)), we can reuse that logic from `remote`.

For copy, you could just change [this](https://github.com/google/go-containerregistry/blob/72597da50ce903895087828b5fc4b918636366f4/pkg/crane/copy.go#L48-L52) to:

```go
if o.platform != nil {
                // If platform is explicitly set, don't copy the whole index, just the appropriate image.
  		if err := copyImage(desc, dstRef, o); err != nil {
			return fmt.Errorf("failed to copy image: %v", err)
		}
} else {
                 if err := copyIndex(desc, dstRef, o); err != nil {
			return fmt.Errorf("failed to copy index: %v", err)
		}
}
```

Since `copyImage` call [`desc.Image`](https://github.com/google/go-containerregistry/blob/72597da50ce903895087828b5fc4b918636366f4/pkg/crane/copy.go#L69), it will do the child matching logic for you.

Fixing `Manifest` and `Digest` would require similar changes.

dsbrng25b commented on pkg/crane/digest.go:
```diff
@@ -16,9 +16,14 @@ package crane
+	if err != nil {
+		return "", err
+	}
+	return "sha256:" + digest.Hex, nil
```
Is there a better way to do this?

jonjohnsonjr replied:
`digest.String()` should work

jonjohnsonjr commented on pkg/crane/get.go:
```diff
@@ -35,11 +40,20 @@ func getImage(r string, opt ...Option) (v1.Image, name.Reference, error) {
+	}
+
+	if o.platform == nil {
+		return desc.ImageIndex()
```
This will fail for anything that's not an index (schema1 or image) -- I'd probably just inline the logic from `getManifest` into `Manifest` and `Digest` -- the awkwardness of trying to fit this into the type system makes it tough.

This would be a lot easier if I hadn't (stupidly) embedded `v1.Descriptor` into `remote.Descriptor`, because we have a name collision between the `Digest` in the struct in the `Digest()` method 😞

dsbrng25b commented:
I moved the logic to the Digest and Manifest function. Does not look super nice, but is probably easier to understand than the logic I used before with the interface.

jonjohnsonjr commented on pkg/crane/manifest.go:
```diff
@@ -14,11 +14,36 @@
 		return nil, err
 	}
-	return desc.Manifest, nil
+	return img.RawManifest()
```
This will still fail for schema 1 images, which we aren't able to satisfy the `v1.Image` interface for. If we keep the behavior the same here, but only try to resolve to an image when platform is specified, I think we can keep everything working, i.e.:

```go
if o.platform != nil {
  img, err := desc.Image()
  if err != nil {
    return nil, err
  }
  return img.RawManifest()
}

return desc.Manifest, nil
```

Sorry for all the back and forth on this, I really appreciate your contribution!

dsbrng25b commented:
Tried to change the logic according to your points in the last review, squashed the commits and rebased them on master.
Since in the meantime crane.Digest was changed and now only uses HEAD I had to change the logic that it still gets the manifest if a platform is specified. Or is there another solution.

jonjohnsonjr commented:
> I had to change the logic that it still gets the manifest if a platform is specified. Or is there another solution?

I think what you've done is the best we can do.

jonjohnsonjr reviewed (approved):
LGTM, thanks!
